tools/redcap2linst.php

```diff
@@ -0,0 +1,166 @@
+    "f:",
+    [
+        "output-dir:",
+        "file:",
+    ]
+);
+
+$inputFile = $options['f'] ?? $options['file'] ?? '';
+$outputDir = $options['output-dir'] ?? '';
+
+if (empty($inputFile) || empty($outputDir)) {
+    showHelp();
+    exit(1);
+}
+if (!is_dir($outputDir)) {
+    fprintf(STDERR, "Output directory $outputDir doesn't exist.\n");
+    exit(1);
+}
+if (!is_writeable($outputDir)) {
+    fprintf(STDERR, "Output directory $outputDir is not writeable.\n");
+    exit(1);
+}
+
+$fp = fopen($inputFile, "r");
+if ($fp === false) {
+    fprintf(STDERR, "Could not open file $inputFile\n");
+    exit(1);
+}
+
+$headers     = fgetcsv($fp);
+$instruments = [];
+while ($row = fgetcsv($fp)) {
+    $inst = $row[1];
+    if (!isset($instruments[$inst])) {
+        $instruments[$inst] = [];
+    }
+    $linstFormat = toLINST($row[3], $row[0], $row[4], $row[5]);
+    if (!empty($linstFormat)) {
+        $instruments[$inst][] = $linstFormat;
+    }
+}
+fclose($fp);
+
+outputFiles($outputDir, $instruments);
+
+/**
+ * Take a single line from the redcap dictionary and returns the
+ * closest LINST equivalent.
+ *
+ * @param string $redcaptype      The type from REDCap
+ * @param string $redcapfieldname The fieldname from REDCap
+ * @param string $redcaplabel     The label from REDCap
+ * @param string $redcapChoices   The choices columnfrom REDCap
+ *
+ * @return string
+ */
+function toLINST(
+    string $redcaptype,
+    string $redcapfieldname,
+    string $redcaplabel,
+    string $redcapChoices,
+) : string {
+    $label = str_replace($redcaplabel, "\n", "<br /><br />");
+    switch ($redcaptype) {
+    case 'text':
+        // text maps directly to LORIS
+        return "text{@}$redcapfieldname{@}$label";
+    case 'descriptive':
+        // descriptive maps to label with no field.
+        return "static{@}{@}$label";
+    case 'radio':
+    case 'dropdown':
+        // Radio or dropdown maps to a select and the options are in the
+        // same format in the dictionary.
+        return "select{@}$redcapfieldname{@}$label{@}"
+            . optionsToLINST($redcapChoices);
+    case 'checkbox':
+        // checkboxes are the same format as radios but allow multiple options,
+        // so map to a selectmultiple instead of a select
+        return "selectmultiple{@}$redcapfieldname{@}$label{@}"
+            . optionsToLINST($redcapChoices);
+    case 'yesno':
+        // Map yes/no fields to dropdowns with yes and no options.
+        return "select{@}$redcapfieldname{@}$label{@}"
+            . "NULL=>''{-}'yes'=>'Yes'{-}'no'=>'No'";
+    case 'calc':
+        // Calc maps to a score field. We create the DB field but don't do the score.
+        return "static{@}$redcapfieldname{@}$label";
+    case 'sql':
+        // The "SQL" data type seems to just be for presentation? I hope?
+        return "";
+    case 'notes':
+        // REDCap calls textareas notes
+        return "textarea{@}$redcapfieldname{@}$label";
+    default:
```

Comment:
could add 'section header' case and map it to linst "header" elements

wondering also if there's any desire to add pages as well? in my other PR, I used the section header in REDCap to split those sections into pages, and then add the header at the top of the page

Reply:
Do you have an example of a section header? Is there a type or are you talking about watching the column and keeping track of when it changes?

Reply:
section header would be like this:

```
page{@}{@}Attention and Working Memory
header{@}{@}Attention and Working Memory
```

whenever there's an entry in "Section Header" column on redcap data dictionary, i add a page and a header element to translate that section, so yes watching the column and keeping track of when its not empty